File: java/Programmers/Lv.2/Programmers_미로탈출.java
import java.util.*;

public class Programmers_미로탈출 {
    static int[] dr = { -1, 0, 1, 0 };
    static int[] dc = { 0, 1, 0, -1 };

    static class Pos {
        int r, c, cnt;

        Pos(int r, int c, int cnt) {
            this.r = r;
            this.c = c;
            this.cnt = cnt;
        }
    }

    static boolean find = false;
    static int answer;

    public int solution(String[] maps) {
        outer: for (int i = 0; i < maps.length; ++i) {
            for (int j = 0; j < maps[i].length(); ++j) {
                char now = maps[i].charAt(j);
                if (now == 'S') {
                    bfs(i, j, maps, 0, false);
                    break outer;
                }
            }
        }
        if (!find)
            answer = -1;
        return answer;
    }

    static public void bfs(int r, int c, String[] maps, int cnt, boolean flag) {
        Queue<Pos> q = new LinkedList<>();
        boolean[][] visited = new boolean[maps.length][maps[0].length()];
        q.offer(new Pos(r, c, cnt));
        visited[r][c] = true;
        while (!q.isEmpty()) {
            Pos now = q.poll();
            if (maps[now.r].charAt(now.c) == 'L' && !flag) {
                bfs(now.r, now.c, maps, now.cnt, true);
                return;
            }
            if (maps[now.r].charAt(now.c) == 'E' && flag) {
                answer = now.cnt;
                find = true;
                return;
            }
            for (int i = 0; i < 4; ++i) {
                int nr = now.r + dr[i];
                int nc = now.c + dc[i];
                if (nr >= maps.length || nc >= maps[0].length() || nr < 0 || nc < 0) {
                    continue;
                }
                if (!visited[nr][nc] && maps[nr].charAt(nc) != 'X') {
                    visited[nr][nc] = true;
                    q.offer(new Pos(nr, nc, now.cnt + 1));
                }
            }
        }

    }
}

File: java/Programmers/Lv.2/Programmers_미로탈출.cpp
#include <string>
#include <vector>
#include <iostream>
#include <queue>

using namespace std;
vector<int> dr = {-1, 0, 1, 0};
vector<int> dc = {0,1, 0,-1};
class Pos{
 public :
    int r;
    int c;
    int cnt;
    Pos(int r, int c, int cnt) {
        this-> r = r;
        this-> c = c;
        this-> cnt = cnt;
    }
};
bool findAns;
int answer;
void bfs(int r, int c, vector<string> maps, int cnt, bool flag){
    queue<Pos> q;
    int visited[maps.size()][maps[0].size()];
    for(int i =0 ;i < maps.size(); ++i){
        for(int j = 0; j < maps[i].size(); ++j){
            visited[i][j]= 0;
        }
    }
    q.push(Pos(r,c,cnt));
    visited[r][c] = 1;
    while(!q.empty()){
        Pos now = q.front();
        q.pop();
        if(maps[now.r].at(now.c)=='L' && !flag){
            bfs(now.r, now.c, maps, now.cnt, true);
            return;
        }
        if(maps[now.r].at(now.c)=='E' && flag){
            answer = now.cnt;
            findAns = true;
            return;
        }
        for(int i = 0; i < 4; ++i){
            int nr = now.r + dr[i];
            int nc = now.c + dc[i];
            if(nr >= maps.size() || nc >= maps[0].size() || nr < 0 || nc < 0) continue;
            if(visited[nr][nc] == 0 && maps[nr].at(nc)!='X'){
                visited[nr][nc] = 1;
                q.push(Pos(nr,nc,now.cnt+1));
            }
        }
    }
}
int solution(vector<string> maps) {
    bool start = false;
    for(int i = 0; i < maps.size(); ++i){
        for(int j = 0; j < maps[i].size(); ++j){
            if(maps[i].at(j)=='S'){
                bfs(i, j, maps, 0, false);
                break;
            }
        }
        if(start) break;
    }
    if(!findAns) answer = -1;
    return answer;
}
